### PR TITLE
cmake: Copy tests to the build tree instead of linking them

### PR DESCRIPTION
**Pull request by hebasto**
This PR enables in-source builds, which currently fail on the staging branch.

As discussed offline, we won't restrict the user, even though CMake discourages in-source builds.

Other features of this approach:
PROS:
1. Less of code. Simpler code.
2. The same code for both POSIX and Windows platforms.
3. It prevents the source tree from being polluted with Python bytecode caches for the test framework when running functional tests. For reference, https://github.com/bitcoin/bitcoin/pull/30533 aims for the same goal for unit tests.

CONS:
1. Approximately 4MB larger build directory. Note that it is smaller than its size just after configuring on the staging branch.

Resolves https://github.com/hebasto/bitcoin/issues/306.

### CODE REVIEW

**Comment by hebasto** [comment]
cc:
- @achow101 @m3dwards who voted for freedom of the user's choice
- @paplorinc as an author of https://github.com/hebasto/bitcoin/issues/306

**Comment by m3dwards** [comment]
I'm getting this on Arm Mac

```shell
➜ cmake .

CMake Error at cmake/module/AddBoostIfNeeded.cmake:72 (message):
  Building test_bitcoin executable requested but
  boost/test/included/unit_test.hpp header not available.
Call Stack (most recent call first):
  CMakeLists.txt:406 (add_boost_if_needed)
```

**Comment by m3dwards** [comment]
ACK af3b04244ef19d484169356a62e92759e1c82b98

Tested on Arm Mac and x86 Linux. Can replicate problems without and can configue/build/test with PR.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -32,17 +32,6 @@ endfunction()
...
-foreach(script ${functional_tests} fuzz/test_runner.py util/rpcauth-test.py util/test_runner.py)
-  if(CMAKE_HOST_WIN32)
-    set(symlink)
-  else()
-    set(symlink SYMBOLIC)
-  endif()
-  file(CREATE_LINK ${CMAKE_CURRENT_SOURCE_DIR}/${script} ${CMAKE_CURRENT_BINARY_DIR}/${script} COPY_ON_ERROR ${symlink})
-endforeach()
-unset(functional_tests)
+file(COPY functional util/rpcauth-test.py DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
```
[the remark this answers is not in the document]
Comment (hebasto) [reply test/CMakeLists.txt:35]: > q: Creating a copy here, removes [bitcoin#30463](https://github.com/bitcoin/bitcoin/pull/30463) again as a cmake requirement?

Correct.

**Comment by fanquake** [comment]
~0. Not sure. After a year of deciding we'll do things the "CMake way", at the last minute we're going to start allowing the thing CMake actively discourages, because of an unintuitive error message? Can't we just provide a more useful error?

The PROs listed here are all pretty weak, and don't point to anything that's actually useful for a developer, doing development. It'd be good to get some example usecases of why someone might want to build in tree, that doesn't (or for some reason can't?) work with out-of-tree builds. As opposed to allowing this just because someone might "want to", and opening us up to also maintaining that can of worms (especially if it's not something we are going to actually use or test).

**Review by TheCharlatan — COMMENTED** [review]
I tend towards NACK here for the downsides brought up already. I also think adding a flag for changing this behaviour and allowing in-tree builds is not a good alternative - especially absent a clear use-case. I also remember reviewing the script linking behaviour and thinking this was a good improvement.

**Comment by hebasto** [comment]
> I tend towards NACK here for the downsides brought up already. I also think adding a flag for changing this behaviour and allowing in-tree builds is not a good alternative - especially absent a clear use-case. I also remember reviewing the script linking behaviour and thinking this was a good improvement.

Please consider an alternative in https://github.com/hebasto/bitcoin/pull/314.

**Comment by achow101** [comment]
Is it possible to allow in source builds while preserving the symlink behavior for out of tree builds? If I do an out of tree build with this PR, changes I make to the functional tests require an additional cmake invocation for those changes to appear in the build. When working on test changes, that's really annoying.

**Comment by hebasto** [comment]
> Is it possible to allow in source builds while preserving the symlink behavior for out of tree builds?

Yes, it is.

> If I do an out of tree build with this PR, changes I make to the functional tests require an additional cmake invocation for those changes to appear in the build. When working on test changes, that's really annoying.

That's true.

**Comment by hebasto** [comment]
Rebased.

**Comment by hebasto** [comment]
> symlinks should be kept for out-of-tree builds, otherwise it seems like a breaking change to dev experience when editing files

I agree with this point.

However, instead of adjusting this PR, I'm going to close in favour of https://github.com/hebasto/bitcoin/pull/314.

The main reason is that in-source builds with CMake are incompatible with Autotools because the `Makefile` files are being overwritten.

We can reconsider this idea after completing the migration to CMake.

**Pull request closed without merging**